Reusing scripts for live resources & no more deploy buttons (#8398)
Azure/azure-sdk-for-js: sdk/appconfiguration/app-configuration/test/README.md

# Testing

To test this project, make sure to first build it properly by following our [building instructions](https://github.com/Azure/azure-sdk-for-js/blob/master/CONTRIBUTING.md#building). Once the project is correctly built, you will be able to run the tests by following the [testing instructions](https://github.com/Azure/azure-sdk-for-js/blob/master/CONTRIBUTING.md#testing).

> The environment variables you will need to properly assign to run the live test of this project are the following:

Reviewer: We should have a line about what Azure resource is needed to run the tests.

```suggestion
To run the live tests, you need to have an [Azure App Configuration account](link to some docs that talk about how to create this account) and set the below environment variables:
```

Reviewer: Use similar format for other packages as well

Author: on it

> - `AZ_CONFIG_CONNECTION`: The connection string of your Azure App Configuration account.

Reviewer: ```suggestion
- `TEST_MODE`: live
- `AZ_CONFIG_CONNECTION`: The connection string of your Azure App Configuration account.
```

Reviewer: Applies for all other packages that have recorded tests being used by default

Author: on it

> - `AZ_CONFIG_ENDPOINT`: The endpoint of your Azure App Configuration account.

Reviewer: @richardpark-msft Can you not infer the endpoint from the connection string?

# Testing

To test this project, make sure to first build it properly by following our [building instructions](https://github.com/Azure/azure-sdk-for-js/blob/master/CONTRIBUTING.md#building). Once the project is correctly built, you will be able to run the tests by following the [testing instructions](https://github.com/Azure/azure-sdk-for-js/blob/master/CONTRIBUTING.md#testing).

To run the live tests, follow the [Integration Testing with live services](https://github.com/Azure/azure-sdk-for-js/blob/master/CONTRIBUTING.md#integration-testing-with-live-services) instructions to automatically set up the required resources with the proper configuration.

> The live tests in this project will use the resources created from the [ARM template](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/appconfiguration/test-resources.json), which defines:

Reviewer: > The live tests in this project will use the resources created from the [ARM template]

The "will use" here tells the reader that there is no other way to run the live tests and that they **have** to use the ARM template. The reader should know that they can chose to manually create the resource or use an already created one.

Also, we have already linked to the testing instructions, so we don't need to again link to the `Integration Testing with live services`. The above 2 paragraphs can be replaced with something like the below:

> You can bring your own Azure resources for the live tests to use or you can use our [script that creates the Azure resources](link to the New-TestResources.md file) using an ARM template. The below Azure resources are used by the tests in this project.

Author: > The "will use" here tells the reader that there is no other way to run the live tests and that they have to use the ARM template. The reader should know that they can chose to manually create the resource or use an already created one.

This is way harder than it seems because some of the tests require non trivial resources to make. The alternative to the ARM template is to go one by one documenting how to create all the resources needed to run live tests.

Author: > Also, we have already linked to the testing instructions, so we don't need to again link to the Integration Testing with live services. The above 2 paragraphs can be replaced with something like the below:

Sounds good!

Author: I have updated these files with all of your current feedback considered. Let me know how it looks! Also let me know if I've missed anything!

- An [Azure App Configuration](https://docs.microsoft.com/en-us/azure/azure-app-configuration/overview) account with minimum configurations.

Set the below environment variables to run the live tests:

- `TEST_MODE`: Should have `live` assigned.
- `AZ_CONFIG_CONNECTION`: The connection string of your Azure App Configuration account.
- `AZ_CONFIG_ENDPOINT`: The endpoint of your Azure App Configuration account.

> The Azure App Configuration client live tests will add, modify and delete configurations on the Azure App Configuration account provided by the given `AZ_CONFIG_ENDPOINT` environment variable.

Reviewer: The "client" does not have live tests, the project does.

```suggestion
The live tests in this project will add, modify and delete configurations on the Azure App Configuration account being used.
```

Author: thank you!

Author: I have updated these files with all of your current feedback considered. Let me know how it looks! Also let me know if I've missed anything!

# Testing

> To run our tests, first install the dependencies by running `npm install` (or `rush install`) at the root of the source code of this library.

Author: I've added this information! Thank you!

With the dependencies installed, you'll be able to run the unit tests and integration tests.

## Unit tests

To run unit tests, go to the root of this library and run `npm run unit-test`.
This will run tests in a purely nodejs environment, and in a browser environment using [Karma](https://karma-runner.github.io/latest/index.html).
These tests aim to reproduce the behavior of our library against remote endpoints through previously recorded HTTP request and responses.
You won't need to have any environment configuration, nor internet access, for these tests to run.

Keep in mind that if you modify the test files, the API shape of this library, or the underlying network calls, the recordings need to be re-generated.
You can read more in the section: [Regenerating recordings](#regenerating-recordings).

## Integration tests

> Our integration tests will run against the live resources, which are determined by the environment variables you provide.

Reviewer 2: The unit-tests can run against the live environment, too, and the integration tests don't run with `TEST_MODE=live` by default, so I think this is a little misleading.

Author: I think we should address this in the test guidelines. These scripts should only do what they're supposed to do. I'll add a comment there.

Author: I've added feedback to myself on this PR: https://github.com/Azure/azure-sdk-for-js/pull/7784

To run the integration tests, you will need to execute `npm run integration-test`.

To generate new resources for your tests, follow the procedure at [Integration Testing with live services](https://github.com/Azure/azure-sdk-for-js/blob/master/CONTRIBUTING.md#integration-testing-with-live-services).

The environment variables needed are:

- `AZ_CONFIG_CONNECTION`: The connection string of your Azure App Configuration account.
- `AZ_CONFIG_ENDPOINT`: The endpoint of your Azure App Configuration account.

Integration tests won't need updated recordings, and they also won't change previous recordings unless specified (see [Regenerating recordings](#regenerating-recordings)).

## Regenerating recordings

If you modify the test files, the API shape of this library, or the underlying network calls, the recordings need to be re-generated before you run the unit tests.

> To re-generate the recordings, set the environment variable `TEST_MODE` to `record` and then run the integration tests. Once the integration tests finish,

Reviewer 2: I usually use `TEST_MODE=record rushx unit-test` to create new recordings. Is there a reason that I should prefer the integration tests? Currently I just assume that `integration` is the command that the test pipelines run and `unit` is the command that I run during development. That would obviously change if they started being different sets of tests, but for now I don't think it makes a difference.

Author: Only the integration test command should reach to live resources. I'll make sure the new test guidelines enforce this. Otherwise it's really misleading 😅 More will be elaborated here: https://github.com/Azure/azure-sdk-for-js/pull/7784

# Testing

> To run our tests, first do a full update of the project by running `rush update`, then rebuild either this package, with `rush build -t @azure/app-configuration` or all packages with `rush build`, then make sure to build the tests with `rush build:test`. After that, you can either run our unit tests, which won't reach to live services, and our integration tests, which will indeed require existing, reachable resources.

Reviewer: > then make sure to build the tests with `rush build:test`

The `test`, `test:node` and `test:browser` scripts do build the test code before running the tests. Any reason we want to call out building the tests specifically here?

> After that, you can either run our unit tests, which won't reach to live services, and our integration tests, which will indeed require existing, reachable resources.

I would imagine a contributor to ever just use the `test`, `test:node` or `test:browser` scripts. Both `unit` and `integration` scripts are for CI. Due to this, I would avoid calling out "unit" and "integration" tests in our readme instructions so as to not mislead the user (because our "unit" script doesnt really run "unit" tests :) ). 

Instead, I would call out that by default, the tests are run in playback mode. If you want to target live services, set TEST_MODE to "live" and set the required env variables.

Author: Thank you! I'll make sure the new approach has this covered.

> ## Run all tests

Reviewer: Assuming we are moving to a `Testing` section in the contributing.md file, I would recommend the below structure

```
### Testing

Run `rushx test` from the root of the package that you want to run tests for. 
This will run tests both in node.js and browser. To target them individually, 
you can run `rushx test:node` and `rushx test:browser`

#### Recorded tests

By default, the npm scripts `test`, `test:node` or `test:browser` run recorded tests.
<<Add the content that you now have under "Unit Tests">>

#### Live tests

By default, the npm scripts `test`, `test:node` or `test:browser` run recorded tests.
To run live tests, set the environment variable TEST_MODE to "live" and ensure you have the right 
Azure resources to test against.
<< Add the content that you now have under "integration tests">>

#### Debugging tests using Visual Studio Code

```

I would skip the below sections you have now. These are generic instructions on how to use mocha
- Run all tests in a single test suite
- Run a single test

Author: Thank you so much! This really helps!

Run `rushx test:node` or `rushx test:browser` from your terminal to run all the tests under the `./test` folder

## Run all tests in a single test suite

Append the `.only` on the `describe` method corresponding to the test suite.

Then run `rushx test:node` or `rushx test:browser` from your terminal.

## Run a single test

Append the `.only` on the `it` method corresponding to the test.

Then run `rushx test:node` or `rushx test:browser` from your terminal.

## Debug tests using Visual Studio Code

- Select the debug configuration `Debug Unit Tests` from the drop down in the debug viewlet.
- Add breakpoints in the test you want to debug. Tip: It is helpful to append the `.only` on the test method if you want only that method to be run.

> - Start Debugging

Reviewer: For debugging using VS Code, the main catch is going to be to ensure you are in the right folder and not root of the repo so that the debug configuration can be picked up. Unless we have one at the root as well which works with all packages. 

I would make the assumption that the reader would know how to work with VS Code and the tip with `.only`. I would keep this short

> Each package has a debug configuration `Debug Unit Tests` that you can use to debug your tests.
Ensure that you have opened the right folder in VS Code so that the corresponding debug configuration can get picked up.

Also, I wouldnt add this section until we first make sure all Track 2 packages do have a working debug configuration.

Author: I will remove this section then. Thank you!

## Unit tests

> To run unit tests, go to the root of this library and run `rushx unit-test`. This will run tests in a purely nodejs environment, and in a browser environment using [Karma](https://karma-runner.github.io/latest/index.html). These tests aim to reproduce the behavior of our library against remote endpoints through previously recorded HTTP request and responses. You won't need to have any environment configuration, nor internet access, for these tests to run.

Author: ```suggestion
To run unit tests, go to the root of this library and run `rushx unit-test`. This will run tests in NodeJS and in a browser (through [Karma](https://karma-runner.github.io/latest/index.html)). These tests aim to reproduce the behavior of our library against remote endpoints through previously recorded HTTP request and responses. You won't need to have any environment configuration, nor internet access, for these tests to run.
```
